MIPS/mip-zkapps.md

```diff
@@ -0,0 +1,961 @@
+until now checked only signatures. After verified in the pool, the proofs are
+assumed to be valid during block creation and don't require checking again,
+similar to signatures for signed commands. Proofs within a transaction and
+across multiple transactions can be batched to make the verification step
+slightly more efficient. Failing batch verification involves additional
+verification to identify the faulty proof or transaction and so, worst case each
+transaction are actually a bit slower.
+Additionally, hashing zkApp transactions impact the pool's performance.
+Care should be taken to hash a transaction only once and use it everywhere in
+the pool and throughout the protocol.
+
+#### Block production
+
+ZkApp transactions in blocks can increase block production time. For example, a
+zkApp transaction with four account updates is equivalent to two payments. So
+essentially with zkApp transactions there can be more accounts to be updated
+(or more updates per account) with the same number of transactions per blocks as
+mainnet is currently at. The overhead mainly comes from additional steps
+involved in updating the scan state which includes witness generation.
+
+#### Block Validation
+
+Block verification now involves additional zkApp proof verification. The
+proofs can be efficiently batch verified since a failing batch means the block
+is simply rejected.
+
+#### Snark workers
+
+A ZkApp transaction is essentially a list of updates to accounts. Generating
+transaction snark for a zkApp transaction involves proving each account
+update and the recursively merging the proofs. Thus proving time of a zkApp
+transaction depends on the number of account updates. Additionally, account
+updates with proofs are more expensive than account updates with signatures or
+no authorization. Because the transaction application logic circuit will
+increase, more memory is required to create transaction proofs. This MIP
+requires that Snark Workers utilize more compute and memory than they had before
+this change.
+
+#### Memory impact
+
+In addition to the impact to Snark Workers, all participating nodes in the
+network must now store larger transactions through different
+subsystems. Blocks will be larger because transactions can be larger. The
+mempool can be tweaked to store fewer max transactions at a time, but other
+components that store blocks must expand in size. This has some effect on
+memory, but initial prototypes confirm node requirements remain reasonable.
+
+#### Limiting zkApp transaction size
+
+There is a "size heuristic" described earlier in this document to ensure the
+number of account updates inside zkApp transactions is limited. There are also
+limits on other parts of a transaction: The memo has a fixed size and events
+and actions are capped to enforce that the network doesn't need to do too much
+work to process a transaction.
+
+## Backwards Compatibility
+
+This MIP is not backwards-compatible; it is an extensive change to the
```

Comment:
```suggestion
This MIP is not backward-compatible; it is an extensive change to the
```

Reply:
I rely on [Google developer documentation style guide](https://developers.google.com/style/word-list#letter-b), where I looked up this term